src/art/art.c

```diff
@@ -1448,118 +1673,377 @@ static void art_node_printf(const art_node_t *node, uint8_t depth) {
+    // This can be done by traversing the list until we find a node with a
+    // `next_free` greater than the index we copied the node from, and inserting
+    // the new index in between. This leads to the following:
+    //     nodes = [x,2,3,5,x]
+    //     first_free = 1
+    uint64_t initial_next_free = next_free;
+    uint64_t current = next_free;
+    while (next_free < from) {
+        current = next_free;
+        next_free =
+            art_node_get_next_free(art, art_to_ref(next_free, typecode));
+    }
+    art_node_set_next_free(art_deref(art, ref), typecode, next_free);
+    if (current < from) {
+        art_node_set_next_free(art_get_node(art, current, typecode), typecode,
+                               from);
+    }
+    art->first_free[typecode] =
+        from < initial_next_free ? from : initial_next_free;
+    return art_to_ref(to, typecode);
+}
+
+/**
+ * Sorts the free lists pointed to by art->first_free in ascending index order.
+ */
+static void art_sort_free_lists(art_t *art) {
+    for (art_typecode_t type = CROARING_ART_LEAF_TYPE;
+         type <= CROARING_ART_NODE256_TYPE; ++type) {
+        bool *free_indices =
+            (bool *)roaring_calloc(art->capacities[type], sizeof(bool));
+
+        for (uint64_t i = art->first_free[type]; i < art->capacities[type];
+             i = art_node_get_next_free(art, art_to_ref(i, type))) {
+            free_indices[i] = true;
+        }
+
+        uint64_t first_free = art->capacities[type];
+        for (uint64_t i = art->capacities[type]; i > 0; --i) {
+            uint64_t index = i - 1;
+            if (free_indices[index]) {
+                art_node_set_next_free(art_get_node(art, index, type), type,
+                                       first_free);
+                first_free = index;
+            }
+        }
+        art->first_free[type] = first_free;
+        roaring_free(free_indices);
+    }
+}
+
+/**
+ * Shrinks all node arrays to `first_free`. Assumes all indices after
+ * `first_free` are unused.
+ */
+static size_t art_shrink_node_arrays(art_t *art) {
+    size_t freed = 0;
+    if (art->first_free[CROARING_ART_LEAF_TYPE] <
+        art->capacities[CROARING_ART_LEAF_TYPE]) {
+        uint64_t new_capacity = art->first_free[CROARING_ART_LEAF_TYPE];
+        art->leaves =
+            roaring_realloc(art->leaves, new_capacity * sizeof(art_leaf_t));
+        freed += (art->capacities[CROARING_ART_LEAF_TYPE] - new_capacity) *
+                 sizeof(art_leaf_t);
+        art->capacities[CROARING_ART_LEAF_TYPE] = new_capacity;
+    }
+    if (art->first_free[CROARING_ART_NODE4_TYPE] <
+        art->capacities[CROARING_ART_NODE4_TYPE]) {
+        uint64_t new_capacity = art->first_free[CROARING_ART_NODE4_TYPE];
+        art->node4s =
+            roaring_realloc(art->node4s, new_capacity * sizeof(art_node4_t));
+        freed += (art->capacities[CROARING_ART_LEAF_TYPE] - new_capacity) *
```

Comment:
Should be `CROARING_ART_NODE4_TYPE`

This all feels very prone to copy-paste issues. Wonder if it'd be worth it to just define a `void *nodes[6]`, and use macros to get access the nodes by type (including the right casts).

And/Or maybe something like

```c
struct {
  size_t first_free;
  size_t capacity;
  union {
    art_leaf_t *leaf;
    art_node4_t *node4;
    art_node16_t *node16;
    art_node48_t *node48;
    art_node256_t *node256;
  }
} nodes[6];
```

Along with a `static const NODE_SIZES[6]`, would allow writing these functions as loops. Dunno if it's worth it or not.

Reply:
Good idea, using a `void *` array seems to work well.